src/main/kotlin/org/dif/peerdid/PeerDIDUtils.kt

```diff
@@ -0,0 +1,149 @@
+@file:JvmName("PeerDIDUtils")
+
+package org.dif.peerdid
+
+import com.google.gson.GsonBuilder
+import com.zman.varint.VarInt
+import io.ipfs.multibase.Base58
+import io.ipfs.multibase.Multibase
+import io.ipfs.multibase.binary.Base64
+import org.dif.model.EncodingType
+import org.dif.model.JSON
+import org.dif.model.PeerDID
+import org.dif.model.PublicKey
+import org.dif.model.PublicKeyType
+import org.dif.model.PublicKeyTypeAgreement
+import org.dif.model.PublicKeyTypeAuthentication
+import java.nio.ByteBuffer
+import java.util.HashMap
+
+internal fun buildDIDDocNumalgo0(peerDID: PeerDID): String {
+    val inceptionKey = peerDID.substring(11)
+    val encodingAlgorithm = peerDID[10]
+
+    if (!isInEncodingTypes(encodingAlgorithm))
+        throw IllegalArgumentException("Unsupported encoding algorithm of key: $encodingAlgorithm")
+
+    val decodedEncnumbasis = decodeEncnumbasis(inceptionKey, peerDID)
+    val gson = GsonBuilder().setPrettyPrinting().disableHtmlEscaping().create()
+    val diddoc = mapOf(
+        "id" to peerDID,
+        "authentication" to gson.toJsonTree(decodedEncnumbasis)
+    )
+    return gson.toJson(diddoc)
+}
+
+internal fun buildDIDDocNumalgo2(peerDID: PeerDID): String {
+    val keys = peerDID.drop(11)
+    val keysList = keys.split(".")
+    var service = ""
+    val keysWithoutPurposeCode = mutableListOf<String>()
+    for (key in keysList) {
+        if (key[0] != 'S') {
+            val transform = key[1]
+            if (!isInEncodingTypes(transform))
+                throw IllegalArgumentException("Unsupported transform part of PeerDID: $transform")
+
+            keysWithoutPurposeCode.add(key.drop(2))
+        } else service = key.drop(1)
+    }
+    val decodedEncnumbasises = keysWithoutPurposeCode.map { key -> decodeEncnumbasis(key, peerDID) }
+    val decodedService = decodeService(service, peerDID)
+
+    val authentication = mutableListOf<Map<String, String>>()
+    val keyAgreement = mutableListOf<Map<String, String>>()
+
+    for (i in decodedEncnumbasises.indices) {
```

Comment:
Once `decodeEncnumbasis` returns a `PublicKey`, we can write the code below much easier:
```
authentication = decodedEncnumbasises.filterIsInstance<PublicKeyAuthentication>()
keyAgreement = decodedEncnumbasises.filterIsInstance<PublicKeyAgreement>()
```